internal/auth/claims.go

```diff
@@ -0,0 +1,87 @@
+package auth
+
+import (
+	"crypto/rsa"
+
+	jwt "github.com/dgrijalva/jwt-go"
+	"github.com/pkg/errors"
+)
+
+const (
+	RoleAdmin = "ADMIN"
+	RoleTODO  = "TODO:MORE_ROLES"
+)
+
+// signingAlg is the agreed-upon algorithm that this application uses to sign
+// and validate tokens.
+//
+// The method used to sign the JWT must be validated to avoid a critical
+// vulnerability:
+// https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/
+var signingAlg = jwt.SigningMethodRS256
+
+// Claims represents the authorization claims transmitted via a JWT.
+type Claims struct {
+	Roles []string `json:"roles"`
+	jwt.StandardClaims
+}
+
+// GenerateToken generates a JWT token string.
+func GenerateToken(key *rsa.PrivateKey, kid string, claims Claims) (string, error) {
+	tkn := jwt.NewWithClaims(signingAlg, claims)
+	tkn.Header["kid"] = kid
+	return tkn.SignedString(key)
+}
+
+// KeyFunc is used to map a JWT key id (kid) to the corresponding public key.
+// USE CASE:
+// Private keys should be rotated. During the transition
+// period, tokens signed with the old and new keys can coexist by looking up
+// the correct public key by key id (kid).
+//
+// Key-id-to-public-key resolution is usually accomplished via a public JWKS
+// endpoint. See https://auth0.com/docs/jwks for more details.
+type KeyFunc func(kid string) (*rsa.PublicKey, error)
+
+// NewParser is the factory function for a Parser.
+func NewParser(kf KeyFunc) *Parser {
+	return &Parser{
+		kf: kf,
+		p: &jwt.Parser{
+			ValidMethods: []string{signingAlg.Name},
+		},
+	}
+}
+
+// Parser wraps jwt.Parser with the ability to fetch keys based on kid.
+type Parser struct {
+	kf KeyFunc
+	p  *jwt.Parser
+}
+
+// ParseClaims from a token string.
+func (p *Parser) ParseClaims(tknStr string) (Claims, error) {
+	var claims Claims
+
+	tkn, err := jwt.ParseWithClaims(tknStr, &claims, func(t *jwt.Token) (interface{}, error) {
```

Comment:
Please put the literal function above and in its own variable `f`. I can never read code like this and find it super clever. Plus, the literal function is so much easier to read when it is on its own.